### DISCUSSION TITLE
custom-shader not having any effect

### DISCUSSION THREAD

**Discussion by maxandersen** [ghostty-org/ghostty#4016] [Q&A] [answered]
I tried setting custom-shader to a bunch of various shader found but no matter what it has zero effect for me running ghostty on osx.

Docs says to look in logs but I see nothing in there related to any warnings/errors about shaders.

I tried using deliberate nonexisting paths and that also does not result in entries in log.

any tips on how to debug/trace why a custom-shader is not loading/working?

**Comment by rhodes-b** [comment]
Docs says to look in logs but I see nothing in there related to any warnings/errors about shaders

if you have `cusom-shader` in your config the logs should output the shader it tries to load

`warning(grid): error loading custom shaders err=error.FileNotFound`
`info(shadertoy): loaded custom shader path=/home/user/workspace/test/ghostty-shaders/negative.glsl`

are two results you should see. Try passing the absolute path to the shader if you havent done that

I'm pretty sure it use to tell you the path it tried to load even if it failed this may have changed and should be changed back if that is the case

**Reply by rhodes-b** [reply]
https://github.com/ghostty-org/ghostty/discussions/3304
this could also be related if it is a real issue

**Reply by maxandersen** [reply]
I've used full paths. I'm pretty sure the config is read because if i put a typo in the key custom-shader ie. Bad-shader then it complains about bad key. But if i set it back to proper name nothing happens.

**Reply by rhodes-b** [reply]
Can you share which shader you are using? Maybe that one specifically doesnt work.

It requires its written in glsl and uses the shadertoy api

**Reply by maxandersen** [reply]
Amongst many I tried bunch from https://github.com/m-ahdal/ghostty-shaders which at least claim they work with ghostty, i.e. https://github.com/m-ahdal/ghostty-shaders/blob/main/bettercrt.glsl

**Reply by maxandersen** [reply]
ok so now i re-tried a bunch of things and seems I need to do a full restart of ghostty then *sometimes* I get impact of shaders.

...still useful if log would show/handle errors somehow?

**Reply by maxandersen** [reply]
yeah so most definitely I need to full restart - not just reload. is that expected? if expected maybe log should print custom-shader not having effect on reload?

**Reply by maxandersen** [reply, marked as answer]
just noticing in the docs at https://ghostty.org/docs/config/reference

"This can be repeated multiple times to load multiple shaders. The shaders will be run in the order they are specified.

Changing this value at runtime and reloading the configuration will only affect new windows, tabs, and splits."

not sure if new or I simply missed it but then its all just as expected. sorry for the noise.

**Comment by ipaqmaster** [comment]
I'm also setting `custom-shader` to various shaders in the `ghostty-shaders` repository after after cloning it locally and the logs acknowledge the shader with lines like `info(shadertoy): loaded custom shader path=/tmp/ghostty-shaders/galaxy.glsl` but nothing happens even if I entirely close and re-launch the shell.

They don't seem to be having any impact.

**Reply by jcollie** [reply]
Depending on your setup, "closing all of the windows" may not actually quit Ghostty, it may be running windowless in the background.

**Reply by ipaqmaster** [reply]
Ok I've just tried `gears-and-belts.glsl` and that _does_ seem to be working.

In that case it seems some of these shaders simply don't work? Or they're being accidentally "covered" by the background color or something. I notice when I set `animated-gradient-shader.glsl` and open a new tab there's one frame where there appears to be some kind of gradient before the terminal draws over it.

**Reply by ipaqmaster** [reply]
Aha, I can see the `animated-gradient-shader.glsl` now after setting `background = #000000` in the default config it generated for me

**Reply by ipaqmaster** [reply]
I can reproduce this by commenting my `#background = #000000`. The default background it's using for me is covering up some of these shaders.
